streamingccl: source cluster producer job in streaming replication

gh-casper opened the pull request:
Introduce a producer job in source cluster that tracks a stream replication's liveness. The stream replication will be stopped if the job has been inactive for a period of time.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71833)
<!-- Reviewable:end -->

miretskiy reviewed (commented):
This is a good start.  

> Reviewed 1 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-:-MmZg6nL6Vz3Ncl6sNAL:busbdsn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper and @miretskiy)

___
*[pkg/ccl/streamingccl/streamingutils/utils.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_4OuEDx9wISG6gRsf:-Mm_4OuEDx9wISG6gRsg:bpl77tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamingutils/utils.go#L24)):*
> ```Go
> 
> func init() {
> 	streamingccl.RegisterStreamAPI("complete_stream_ingestion_job",
> ```

I'm not going to review API changes -- see comment in the producer_job test.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_-RlfEGfqV6koQQtE:-Mm_-RlfEGfqV6koQQtF:bh0tuit) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job.go#L70)):*
> ```Go
> 
> 	t := timeutil.NewTimer()
> 	t.Reset(20 * time.Millisecond) // change it a cluster settings or a knob
> ```

That's way too fast for default.  I'd say once a minute is fine;.  I would also pull that into a constant
defaultTimeoutCheckingFrequency.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_-cTU7p-qnbquuyrh:-Mm_-cTU7p-qnbquuyri:b-uciiff) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job.go#L77)):*
> ```Go
> 		case <-t.C:
> 			t.Read = true
> 			t.Reset(1 * time.Millisecond)
> ```

I suspect you're doing it because of the tests.  Let's not.
I would suggest we add a `TimeSource` member to the producerJobResumer; initialize that timesource to "real" time.
Create timer with defaultTimeoutFrequency.  You should be able to manipulate time in your tests, so that you don't need to hard code such low freqencies.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_08k9E6y5d19k2UQ7:-Mm_08k9E6y5d19k2UQ8:b4z686z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job.go#L82)):*
> ```Go
> 				return err
> 			} else if isTimedOut(j) {
> 				return errors.Errorf("stream %d timed out", p.job.ID())
> ```

let's be more specific "replication ststream %d timed out".
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZgMCID6V6sQVFwcfG:-MmZgMCID6V6sQVFwcfH:b-wgyzyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L31)):*
> ```Go
> 		ServerArgs: base.TestServerArgs{
> 			Knobs: base.TestingKnobs{
> 				JobsTestingKnobs: &jobs.TestingKnobs{
> ```

use `jobs.NewTestingKnobsWithShortIntervals()`
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_0zVQ69UVglvn3KBe:-Mm_0zVQ69UVglvn3KBf:b-oi7shn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L49)):*
> ```Go
> 	t.Run("inactive-job-eventually-fails", func(t *testing.T) {
> 		tenantID := roachpb.MakeTenantID(10)
> 		rows := sql.QueryStr(t, "SELECT crdb_internal.init_stream($1, $2)", tenantID.ToUint64(), "2s")
> ```

Even though it appears that this comment is for this test -- it's actually about this whole PR.
I would love to see this PR split into multiple commits; The only thing I want to see in the first commit is the new
resumer implementation.  The rest (the API changes, the crdb_internal, etc, etc) should go into a separate commit.

You might wonder, how could you test this new resumer if you don't have `crdb_internal.init_stream`?  
That's actually quite simple: you can create startable job directly (see CreateStartableJobWithTxn, and examples how to use it in changefeed_stmt).
You basically going to give it a JobRecord -- which in this case is very, very simple ;
You start the job.  Then you proceed to execute the rest of the test.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_2C_RA5w8bKobuuaJ:-Mm_2C_RA5w8bKobuuaK:b-tsl7e2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L54)):*
> ```Go
> 		sql.CheckQueryResults(t, jobsQuery, [][]string{{"running"}})
> 		sql.SucceedsSoonDuration = time.Minute
> 		sql.CheckQueryResultsRetry(t, jobsQuery, [][]string{{"failed"}})
> ```

I think I understand why you would need to have duration set to 1 minute or more.  But, I don't think we should do that either.
See my comment about injecting a timesource into job resumer. To use that injection,
you could add something along the following lines above, where you initialize JobsTestingKnobs:

```
 // Start the server as you do now...

// Manual time control for this test...
ts := timeutil.NewManualTime(timeutil.Now())
registry := server.JobRegistry().(*jobs.Registry)
registry.TestingResumerCreationKnobs = map[jobspb.Type]func(raw jobs.Resumer) jobs.Resumer{
jobspb.TypeStreamReplication: func(raw jobs.Resumer) jobs.Resumer {
		r := raw.(*streamReplicationResumer)
                r.timeSource = ts
}
``` 

So, now, when you CreateStartable job above, that job will use your test controlled timesource.  And with that timesource, you can time travel.
For example, you can advance the time past the initial expiration time you have set so that your job resumer wakes up immediately, loads 
progress and sees that it expired -- all pretty much immediately.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_24C35DD1z3YO45bw:-Mm_24C35DD1z3YO45bx:b4g2os9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L63)):*
> ```Go
> 		row.Scan(&streamID)
> 
> 		testDuration := 30 * time.Second
> ```

If this test is running for 30 seconds -- we need to fix it.  This is a unit test, and should run reasonably fast.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_4AND4WfXVbFXEStI:-Mm_4AND4WfXVbFXEStJ:bmo1ocp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L78)):*
> ```Go
> 			}
> 		}
> 		go progressUpdater()
> ```

I suspect this test is flaky.  Same thing as above -- inject custom time handler.
You don't need to run go routine to update progress with real time -- time travel is faster.
___
*[pkg/jobs/registry.go, line 1234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_4KBw47eB74JeeTd6:-Mm_4KBw47eB74JeeTd7:b-ipiw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/registry.go#L1234)):*
> ```Go
> 			return sErr
> 		}
> 		// reverting
> ```

stray comment?
___
*[pkg/jobs/jobspb/jobs.proto, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZydfuAjwfw-okyj89:-MmZydfuAjwfw-okyj8A:b-2php02) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/jobspb/jobs.proto#L108)):*
> ```protobuf
> 
>   // Key span we are replicating
>   roachpb.Span span = 2;
> ```

I would include either the tennant ID, or span.  not both.
Come to think of it, I think it would be better to just store
the list of spans to stream -- much like it's done for changefeed jobs
(`repeated roachpb.Span spans = 1`)

@dt wdyt?
___
*[pkg/jobs/jobspb/jobs.proto, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZz42zAH676OYsKHqd:-MmZz42zAH676OYsKHqe:bqhds02) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/jobspb/jobs.proto#L111)):*
> ```protobuf
> 
>   // Timeout after which we clean up the job
>   int64 stream_timeout = 3 [(gogoproto.casttype)="time.Duration"];
> ```

I don't think we need this field.  When you create replication stream, you will set progress to expire "in the future".
The amount of time you increment timeout during heartbeat doesn't need to be configurable (at least for now)... We could have cluster setting perhaps.
___
*[pkg/jobs/jobspb/jobs.proto, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZzJtHB57L2nKDi9VG:-MmZzJtHB57L2nKDi9VH:b922xzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/jobspb/jobs.proto#L116)):*
> ```protobuf
> message StreamReplicationProgress {
>   // Last active timestamp in micros
>   int64 last_active = 1;
> ```

See comment above --You can rename it as expiration_time.
You should also change the expiration time to be an hlc.Timestamp,  or you could `import "google/protobuf/timestamp.proto"` and use e.g.
`google.protobuf.Timestamp expiration = 1 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];`


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper and @miretskiy)

___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_0zVQ69UVglvn3KBe:-Mm_4a3G06Hsk4-Wj7jc:bkx5q6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Even though it appears that this comment is for this test -- it's actually about this whole PR.
I would love to see this PR split into multiple commits; The only thing I want to see in the first commit is the new
resumer implementation.  The rest (the API changes, the crdb_internal, etc, etc) should go into a separate commit.

You might wonder, how could you test this new resumer if you don't have `crdb_internal.init_stream`?  
That's actually quite simple: you can create startable job directly (see CreateStartableJobWithTxn, and examples how to use it in changefeed_stmt).
You basically going to give it a JobRecord -- which in this case is very, very simple ;
You start the job.  Then you proceed to execute the rest of the test.
</blockquote></details>

To add to the above point: I'm not saying it's a bad idea to test via sql statement.  But, you can add those tests later, and also, you're not really testing the jobs system -- you're testing the functionality of your resumer -- so, you create this resumer (via sql or directly) is probably less important.


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @shermanCRL)

___
*[pkg/ccl/streamingccl/streamingutils/utils.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_4OuEDx9wISG6gRsf:-Mn24y7x5l0EKno5UFx1:b-vsi2kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamingutils/utils.go#L24)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not going to review API changes -- see comment in the producer_job test.
</blockquote></details>

I split it into two commits. Please review the first commit for API changes.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_-RlfEGfqV6koQQtE:-Mn23na_2cghwFd6F-Ba:b-lfj6nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job.go#L70)):*
> defaultTimeoutFrequency
Done. Created DefaultJobLivenessTrackingFrequency.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_-cTU7p-qnbquuyrh:-Mn23uyl-fqbNTs-VLot:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job.go#L77)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I suspect you're doing it because of the tests.  Let's not.
I would suggest we add a `TimeSource` member to the producerJobResumer; initialize that timesource to "real" time.
Create timer with defaultTimeoutFrequency.  You should be able to manipulate time in your tests, so that you don't need to hard code such low freqencies.
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_08k9E6y5d19k2UQ7:-Mn241G3EYFa8JLkBanY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job.go#L82)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

let's be more specific "replication ststream %d timed out".
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZgMCID6V6sQVFwcfG:-Mn24AKw30CxLAVEuPzk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L31)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

use `jobs.NewTestingKnobsWithShortIntervals()`
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_0zVQ69UVglvn3KBe:-Mn24DXl4VE2CsUQLDSJ:bb9p4eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

To add to the above point: I'm not saying it's a bad idea to test via sql statement.  But, you can add those tests later, and also, you're not really testing the jobs system -- you're testing the functionality of your resumer -- so, you create this resumer (via sql or directly) is probably less important.
</blockquote></details>

Done. Thanks for the tips.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_2C_RA5w8bKobuuaJ:-Mn24_vU7ywCwBiG1hyL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L54)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think I understand why you would need to have duration set to 1 minute or more.  But, I don't think we should do that either.
See my comment about injecting a timesource into job resumer. To use that injection,
you could add something along the following lines above, where you initialize JobsTestingKnobs:

```
 // Start the server as you do now...

// Manual time control for this test...
ts := timeutil.NewManualTime(timeutil.Now())
registry := server.JobRegistry().(*jobs.Registry)
registry.TestingResumerCreationKnobs = map[jobspb.Type]func(raw jobs.Resumer) jobs.Resumer{
jobspb.TypeStreamReplication: func(raw jobs.Resumer) jobs.Resumer {
		r := raw.(*streamReplicationResumer)
                r.timeSource = ts
}
``` 

So, now, when you CreateStartable job above, that job will use your test controlled timesource.  And with that timesource, you can time travel.
For example, you can advance the time past the initial expiration time you have set so that your job resumer wakes up immediately, loads 
progress and sees that it expired -- all pretty much immediately.
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_24C35DD1z3YO45bw:-Mn24Mi607lG593oqJH_:b-zaqwn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L63)):*
>  30 se
Done. Changed it to manual timesouce based test.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_4AND4WfXVbFXEStI:-Mn24dHA4Wk1okTrsLQ3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L78)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I suspect this test is flaky.  Same thing as above -- inject custom time handler.
You don't need to run go routine to update progress with real time -- time travel is faster.
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/jobs.proto, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZydfuAjwfw-okyj89:-Mn24g107yE5uUdEL0fS:b-xli5el) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/jobspb/jobs.proto#L108)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would include either the tennant ID, or span.  not both.
Come to think of it, I think it would be better to just store
the list of spans to stream -- much like it's done for changefeed jobs
(`repeated roachpb.Span spans = 1`)

@dt wdyt?
</blockquote></details>

Done. Changed it to be repeated spans.
___
*[pkg/jobs/jobspb/jobs.proto, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZz42zAH676OYsKHqd:-Mn258BP9etua1fJTZGp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/jobspb/jobs.proto#L111)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't think we need this field.  When you create replication stream, you will set progress to expire "in the future".
The amount of time you increment timeout during heartbeat doesn't need to be configurable (at least for now)... We could have cluster setting perhaps.
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/jobs.proto, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MmZzJtHB57L2nKDi9VG:-Mn24oCkEYQSBYl9Rmjg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/jobspb/jobs.proto#L116)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

See comment above --You can rename it as expiration_time.
You should also change the expiration time to be an hlc.Timestamp,  or you could `import "google/protobuf/timestamp.proto"` and use e.g.
`google.protobuf.Timestamp expiration = 1 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];`
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 1234 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mm_4KBw47eB74JeeTd6:-Mn24rQ5Bhm9C71gfEuN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e62608287fc39b2ba80e4c8c529be8c8f4ddadf/pkg/jobs/registry.go#L1234)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

stray comment?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 2 of 12 files at r2, 8 of 20 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, and @shermanCRL)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DIAwAsMK85DToeFV:-Mn2DIAwAsMK85DToeFW:b-ihmze9) (no related file):*
Let's pull job related code into separate PR.  None of the API stuff.
Let's just get the producer job + test checked in -- it's pretty close.
Let's focus on the rest of those changes in a separate PR.

---
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 30 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Kj4337VS0W15aGts:-Mn2Kj4337VS0W15aGtt:b-umptdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L30)):*
> ```Go
> )
> 
> func producerJob(
> ```

nit: perhaps `makeProducerJobRecord` might be a better name?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2OUBAEtLL-2HeCP6m:-Mn2OUBAEtLL-2HeCP6n:bk3bukm) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L31)):*
> ```Go
> 
> func producerJob(
> 	registry *jobs.Registry, tenantID uint64, timeout time.Duration, username security.SQLUsername,
> ```

I suspect timeout should probably be a setting.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Kqyk9aX3nRhiAUU1:-Mn2Kqyk9aX3nRhiAUU2:b-o0kvaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L42)):*
> ```Go
> 		},
> 		Progress: jobspb.StreamReplicationProgress{
> 			Expiration: timeutil.Now().Add(timeout)},
> ```

nit: let's take `}` to the new line
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2N7ck2hWLhEkIRCfo:-Mn2N7ck2hWLhEkIRCfp:bdsyki3) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L73)):*
> ```Go
> 	isTimedOut := func(job *jobs.Job) bool {
> 		progress := p.job.Progress()
> 		return progress.GetStreamReplication().Expiration.Before(p.timeSource.Now())
> ```

I'm not sure where is the `GetStreamReplication()` function defined?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2NGKd1tttaMnoj75u:-Mn2NGKd1tttaMnoj75v:b-9707hn) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L90)):*
> ```Go
> 			if j, err := execCfg.JobRegistry.LoadJob(ctx, p.job.ID()); err != nil {
> 				return err
> 			} else if isTimedOut(j) {
> ```

our style guide suggests not to have "else if" -- just `if err { return }` instead.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 96 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Nnhi4X8r89g3iArw:-Mn2Nnhi4X8r89g3iArx:b-7iuqmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L96)):*
> ```Go
> 	}
> }
> ```

nit: `// OnFailOrCancel implements jobs.Resumer interface`
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2NUtF07vNuyTDpF-2:-Mn2NUtF07vNuyTDpF-3:bpovfsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L107)):*
> ```Go
> 			jobID, err := doInitStream(evalCtx, evalCtx.Txn, tenantID)
> 			return tree.NewDInt(tree.DInt(jobID)), err
> 		})
> ```

Could you pull out any changes related to init_stream and stream API into separate PR?
Let's keep job related stuff only.  
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 80 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2PLJB6XkQcgg4bqoK:-Mn2PLJB6XkQcgg4bqoL:b-6jdn3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L80)):*
> ```
> 		if err := startJob(jobID, jr); err != nil {
> 			t.Fatal(err)
> 		}
> ```


Shorter way to do the above check.  You can do this throughout this test.

```
require.NoError(t, startJob(jobID, jr))
```
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Po9X89O6ce5cwEy1:-Mn2Po9X89O6ce5cwEy2:b-uj1w67) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L83)):*
> ```Go
> 
> 		jobsQuery := fmt.Sprintf("SELECT status FROM system.jobs WHERE id = %d", jobID)
> 		sql.SucceedsSoonDuration = 1 * time.Second
> ```

This might be too aggressive.  I probably wouldn't bother resetting this; or if I did I'd try 10-30 seconds range.
You have to check if this test is not flaky when running under `-race`.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2PZewD9qrOHQKRXq6:-Mn2PZewD9qrOHQKRXq7:bo381bg) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L116)):*
> ```Go
> 		expiration := jr.Progress.(jobspb.StreamReplicationProgress).Expiration
> 
> 		// Set the initial time to be after the timeout
> ```

did you mean before?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Peqo91rE4xdHPW7W:-Mn2Peqo91rE4xdHPW7X:bgkzp08) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L121)):*
> ```Go
> 			t.Fatal(err)
> 		}
> 		sql.CheckQueryResults(t, jobsQuery, [][]string{{"running"}})
> ```

do you need to use CheckQueryResultsRetry?  Yes, you started the job above, but that doesn't mean the transition to running happens immediately.
___
*[pkg/sql/conn_executor.go, line 2441 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2De2z0GLe4tP_5wJg:-Mn2De2z0GLe4tP_5wJh:blntand) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/conn_executor.go#L2441)):*
> ```Go
> 		}
> 		return security.SQLUsername{}
> 	}
> ```

not sure why your PR needs to touch conn_executor.
___
*[pkg/sql/planner.go, line 599 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DjjX8ECxQlFKP0yp:-Mn2DjjX8ECxQlFKP0yq:bzdcaxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/planner.go#L599)):*
> ```Go
> func (p *planner) JobRegistry() interface{} {
> 	return p.ExecCfg().JobRegistry
> }
> ```

Let's not increase planner API -- it's already large enough and it doesn't save you that much in terms of typing.


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @shermanCRL)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DIAwAsMK85DToeFV:-Mn3hs848NfqEav93N3d:b78jlwk) (no related file):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Let's pull job related code into separate PR.  None of the API stuff.
Let's just get the producer job + test checked in -- it's pretty close.
Let's focus on the rest of those changes in a separate PR.
</blockquote></details>

Will create another PR for API changes. This PR will keep the same for now. For this PR, please focus on the non-API changes. Once that one got merged, I will rebase this one on that. We can save quite some time with this approach.

---
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2OUBAEtLL-2HeCP6m:-Mn3f2RH-rpWmDEx_UZ-:b-uq8nz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L31)):*
> timeout
It is indeed defined as a cluster setting StreamReplicationJobLivenessTimeout.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Kqyk9aX3nRhiAUU1:-Mn3fM2o0jRzPz4ikklB:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L42)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: let's take `}` to the new line
</blockquote></details>

Done
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2N7ck2hWLhEkIRCfo:-Mn3fZKd4WEsJ4NHlZG8:bxz09by) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L73)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not sure where is the `GetStreamReplication()` function defined?
</blockquote></details>

This is generated code in jobs.pb.go
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2NGKd1tttaMnoj75u:-Mn3gD9E95CIL_ryDvPw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L90)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

our style guide suggests not to have "else if" -- just `if err { return }` instead.
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 96 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Nnhi4X8r89g3iArw:-Mn3gJLE6tylB2xzKo6F:b-grekwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L96)):*
> OnFailOrCancel implements jobs.Resumer interface
Done
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2NUtF07vNuyTDpF-2:-Mn3h12t8571YFVaCEif:b78jlwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job.go#L107)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Could you pull out any changes related to init_stream and stream API into separate PR?
Let's keep job related stuff only.  
</blockquote></details>

Will create another PR for API changes. This PR will keep the same for now. For this PR, please focus on the non-API changes. Once that one got merged, I will rebase this one on that. We can save quite some time with this approach.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 80 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2PLJB6XkQcgg4bqoK:-Mn3gYgICoizS6sECrTU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L80)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 		if err := startJob(jobID, jr); err != nil {
> 			t.Fatal(err)
> 		}
> ```


Shorter way to do the above check.  You can do this throughout this test.

```
require.NoError(t, startJob(jobID, jr))
```

</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 83 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Po9X89O6ce5cwEy1:-Mn3gbL1DLz-gWXzZLYO:bfqah2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L83)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This might be too aggressive.  I probably wouldn't bother resetting this; or if I did I'd try 10-30 seconds range.
You have to check if this test is not flaky when running under `-race`.
</blockquote></details>

Now I refactored the test to be based on coordinated timer to minimize the flakiness. 
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2PZewD9qrOHQKRXq6:-Mn3gojy7R-VMWmpv93W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L116)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

did you mean before?
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job_test.go, line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2Peqo91rE4xdHPW7W:-Mn3gstz82N26rJY6wPu:b-6tzj72) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/ccl/streamingccl/streamproducer/producer_job_test.go#L121)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do you need to use CheckQueryResultsRetry?  Yes, you started the job above, but that doesn't mean the transition to running happens immediately.
</blockquote></details>

Switched to coordinated timer based test. Please take another look.
___
*[pkg/sql/conn_executor.go, line 2441 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2De2z0GLe4tP_5wJg:-Mn3cOsp7k40Dq_B0Y2b:b-uspeni) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/conn_executor.go#L2441)):*
> conn_executor
The init_stream API needs to use extendedEvalContext object to access information like username which can only be retrieved through planner.User() here.
___
*[pkg/sql/planner.go, line 599 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DjjX8ECxQlFKP0yp:-Mn3dU1p38Xk5Ji41lk_:b-cbb39t) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/planner.go#L599)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Let's not increase planner API -- it's already large enough and it doesn't save you that much in terms of typing.
</blockquote></details>

We can't use types in jobs package in this file because it will create circular dependency. There are lots of interfaces, like PrivilegedAccessor, defined in eval.go to serve the same purpose: avoiding circular dependency. 

PrivilegedAccessor: https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/sem/tree/eval.go#L3302-L3305


<!-- Sent from Reviewable.io -->

gh-casper committed:
streamingccl: source cluster producer job in streaming replication

Introduce a producer job in source cluster that tracks a stream
replication's liveness. The stream replication will be stopped if
 the job has been inactive for a peried of time.

Release note: none

gh-casper reviewed (commented):
Sorry for the misunderstanding. I changed this PR to be single commit and job changes only.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-:-MnMoeVkDVT4QbuzJx3B:b-1p7q3g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @shermanCRL)



<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 1 of 20 files at r3, 5 of 18 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, and @shermanCRL)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DIAwAsMK85DToeFV:-MnNUHD0-a4mQkfSsc92:b-ulv4on) (no related file):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Will create another PR for API changes. This PR will keep the same for now. For this PR, please focus on the non-API changes. Once that one got merged, I will rebase this one on that. We can save quite some time with this approach.
</blockquote></details>

This is very very close to being good to merge.  I have few more nits, and few requests to revert some parts of the code
I don't believe are needed.

Another request: please update PR description on git hub.  

---
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNS6fP9RqASEoWtmIE:-MnNS6fP9RqASEoWtmIF:b3x3wkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/ccl/streamingccl/streamproducer/producer_job.go#L51)):*
> ```Go
> // 1. Tracks the liveness of the replication stream consumption
> // 2. TODO(casper): Updates the protected timestamp for spans being replicated
> func doInitStream(evalCtx *tree.EvalContext, txn *kv.Txn, tenantID uint64) (jobspb.JobID, error) {
> ```

nit: perhaps rename startReplicationStreamJob?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNTw0v2Y6zOt3Jx6Pv:-MnNTw0v2Y6zOt3Jx6Pw:bfq47qu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/ccl/streamingccl/streamproducer/producer_job.go#L52)):*
> ```Go
> // 2. TODO(casper): Updates the protected timestamp for spans being replicated
> func doInitStream(evalCtx *tree.EvalContext, txn *kv.Txn, tenantID uint64) (jobspb.JobID, error) {
> 	registry := evalCtx.ExecConfigAccessor.JobRegistry().(*jobs.Registry)
> ```

I'm not sure you need to pass eval context here -- you might need it down the line, but right now, when looking 
at this PR, I'd say you can just pass in job registry directly?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 56 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNRzB_BXQoFTqK0je4:-MnNRzB_BXQoFTqK0je5:b-gad9ku) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/ccl/streamingccl/streamproducer/producer_job.go#L56)):*
> ```
> registry.CreateStartableJobWithTxn(ctx, &sj, jobID, txn, jr)
> ```

any particular reason we are using adoptable job here as opposed to registry.CreateStartableJobWithTxn(ctx, &sj, jobID, txn, jr)
like in the test?  
___
*[pkg/sql/conn_executor.go, line 2441 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2De2z0GLe4tP_5wJg:-MnNT3Eg7Z7PSovhlaoP:b8i5b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/conn_executor.go#L2441)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

> conn_executor
The init_stream API needs to use extendedEvalContext object to access information like username which can only be retrieved through planner.User() here.
</blockquote></details>

I see.  I understand now why you need this, but still, let's revert this change.
The reasons is that replication stream creation must be restricted to root only.
so, somewhere in your code (probably the caller of the init stream), will do something like:
```
if err := p.RequireAdminRole(ctx, "stream-replication"); err != nil {
		return err
	}
```

So, the user name in your job record is pretty much always `security.RootUserName()`.
___
*[pkg/sql/planner.go, line 599 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DjjX8ECxQlFKP0yp:-MnNTbAm-3Ukqm3ZvVzm:b-yqunjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/planner.go#L599)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

We can't use types in jobs package in this file because it will create circular dependency. There are lots of interfaces, like PrivilegedAccessor, defined in eval.go to serve the same purpose: avoiding circular dependency. 

PrivilegedAccessor: https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/sem/tree/eval.go#L3302-L3305
</blockquote></details>

I see.. I still question if this change belongs with *this* PR.  I think this change is not needed to make this PR work?
Correct me if I'm wrong.
___
*[pkg/sql/sem/tree/eval.go, line 3239 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNU6O_6gdnQXfSrsWF:-MnNU6O_6gdnQXfSrsWG:bo34vob) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/sql/sem/tree/eval.go#L3239)):*
> ```Go
> 	// JobRegistry returns jobs.Registry from ExecutorConfig
> 	JobRegistry() interface{}
> }
> ```

I think you should revert this change as part of this PR.  Same reason as above: you might as well will have to add
something like this later when you add crdb_internal functions, but for this PR it is not needed.


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @shermanCRL)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DIAwAsMK85DToeFV:-MnRp0FA-Ei_nF5_CB-X:b-896fix) (no related file):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This is very very close to being good to merge.  I have few more nits, and few requests to revert some parts of the code
I don't believe are needed.

Another request: please update PR description on git hub.  

</blockquote></details>

Done.

---
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNS6fP9RqASEoWtmIE:-MnRoJD8-w-znGUEzp1t:b8vns1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/ccl/streamingccl/streamproducer/producer_job.go#L51)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: perhaps rename startReplicationStreamJob?
</blockquote></details>

Removed since it's invoked by init_stream.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNTw0v2Y6zOt3Jx6Pv:-MnRoR3a5XztQiHwAJAo:bp0o9zs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/ccl/streamingccl/streamproducer/producer_job.go#L52)):*
> Removed since it's invoked by init_stream.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 56 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNRzB_BXQoFTqK0je4:-MnRo5_cBhsbm6bKRI9C:b2i80ca) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/ccl/streamingccl/streamproducer/producer_job.go#L56)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> registry.CreateStartableJobWithTxn(ctx, &sj, jobID, txn, jr)
> ```

any particular reason we are using adoptable job here as opposed to registry.CreateStartableJobWithTxn(ctx, &sj, jobID, txn, jr)
like in the test?  
</blockquote></details>

Now I changed it to adoptable job since we used short intervals in the test.
___
*[pkg/sql/sem/tree/eval.go, line 3239 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-MnNU6O_6gdnQXfSrsWF:-MnRoePo3Xs6c4Y3NhEc:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea8aa23286ef62bf6d113655d4a068163fc964b6/pkg/sql/sem/tree/eval.go#L3239)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think you should revert this change as part of this PR.  Same reason as above: you might as well will have to add
something like this later when you add crdb_internal functions, but for this PR it is not needed.
</blockquote></details>

Removed.
___
*[pkg/sql/conn_executor.go, line 2441 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2De2z0GLe4tP_5wJg:-MnRol2S7erSkBJhkL9P:bmi66k2) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/conn_executor.go#L2441)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I see.  I understand now why you need this, but still, let's revert this change.
The reasons is that replication stream creation must be restricted to root only.
so, somewhere in your code (probably the caller of the init stream), will do something like:
```
if err := p.RequireAdminRole(ctx, "stream-replication"); err != nil {
		return err
	}
```

So, the user name in your job record is pretty much always `security.RootUserName()`.
</blockquote></details>

Okay. Will do this check in next PR.
___
*[pkg/sql/planner.go, line 599 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DjjX8ECxQlFKP0yp:-MnRouvt6A4NNLh6oaCw:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/919770cb6cc24386cbc9faa4744583ad64c879c1/pkg/sql/planner.go#L599)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I see.. I still question if this change belongs with *this* PR.  I think this change is not needed to make this PR work?
Correct me if I'm wrong.
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 20 files at r3, 11 of 18 files at r5, 6 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71833)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @shermanCRL)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/71833#-Mn2DIAwAsMK85DToeFV:-MnRv3hu8Ky_fOCPd4SL:b2us8eo) (no related file):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks for sticking w/ me.  This change is <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
I don't know if @dt might have some additional comments, but it's good to go from my perspective.

---


<!-- Sent from Reviewable.io -->

gh-casper commented:
bors r+

gh-casper commented:
bors r+
